Make CallOptions and AbstractStub use Deadline.

buchgr opened the pull request:
- Made CallOptions use the Deadline type instead of a
  long to represent a deadline.
- Added new methods CallOptions.withDeadline(Deadline) and
  AbstractStub.withDeadline(Deadline). The methods are
  marked experimental, as the Deadline class is marked
  experimental. These methods are meant to replace
  CallOptions.withDeadlineNanoTime(Long) and
  AbstractStub.withDeadlineNanoTime(Long), which have
  been deprecated.
- Updated CallOptions.toString() to include all fields.

buchgr commented:
PTAL @ejona86 @louiscryan

buchgr commented on core/src/main/java/io/grpc/CallOptions.java:
```diff
@@ -103,10 +118,12 @@ public CallOptions withCompression(@Nullable String compressorName) {
-    newOptions.deadlineNanoTime = deadlineNanoTime;
-    return newOptions;
+    Deadline deadline = deadlineNanoTime != null
+        ? Deadline.after(deadlineNanoTime - System.nanoTime(), TimeUnit.NANOSECONDS)
```
is the potential for inaccuracy here a problem? Same question for `getDeadlineNanoTime()`.

ejona86 replied:
The inaccuracy doesn't bother me. That sort of thing is done every time you call ScheduledExecutor.schedule. This is also deprecated, so I'm not bothered by some minor inaccuracy.

buchgr replied:
@ejona86 but `getDeadlineNanoTime()` now returns `System.nanoTime()` if the deadline is expired, but it would return some value less than `System.nanoTime()` before.

Depending on how people use these methods (e.g. to compare values), it might break their code i.e. one would expect the below to always be true
```java
opts.withDeadlineNanoTime(nanos).getDeadlineNanoTime() == nanos
```

We could retain the original behaviour by making the `Deadline` constructor package private and introduce a package private method `Deadline.deadlineNanos()`, which we immediately deprecate as well. WDYT?

ejona86 replied:
The other option is to change Deadline.timeRemaining() to return negative numbers. That would not result in nanos being equal, but it would be close. It does seem it could be useful to know how far in the past a deadline expired, which seems like an argument for returning negative numbers from timeRemaining(). The argument against returning negative numbers is applications may be buggy in their handling of it.

buchgr replied:
@ejona86 yep :). any preference? I just wanna make sure this change doesn't break any applications. //cc @carl-mastrangelo

ejona86 replied:
@buchgr, returning negative numbers came up in the initial commit of `timeRemaining()`, but @louiscryan chose not to return negatives. I talked with @louiscryan about what you are experiencing, and we're both in favor of having `timeRemaining()` return negatives. So let's go that way. It won't be 100% accurate, but the edge case is racy no matter how you look at it.

carl-mastrangelo replied:
+1 (-1?) to negative numbers.

buchgr commented on core/src/main/java/io/grpc/CallOptions.java:
```diff
@@ -103,10 +118,12 @@ public CallOptions withCompression(@Nullable String compressorName) {
    *                         {@code null} for unsetting the deadline.
    */
+  @Deprecated
   public CallOptions withDeadlineNanoTime(@Nullable Long deadlineNanoTime) {
```
I deprecated the `*NanoTime(...)` methods, as they haven't been marked experimental.

carl-mastrangelo replied:
FWIW, they can probably never go away.

Methods marked deprecated need to point to the correct way to do something in their comments.

lukaszx0 replied:
+1 to what @carl-mastrangelo said. Let's add `@Depreacted` javadoc and point to `withDeadline` ?

ejona86 commented on core/src/main/java/io/grpc/Deadline.java:
```diff
@@ -139,4 +139,9 @@ public long timeRemaining(TimeUnit unit) {
+
+  @Override
+  public String toString() {
+    return String.valueOf(deadlineNanos);
```
Doing "X ns from now" may not be a bad idea. Just the nanos is useless without a reference to the clock.

buchgr replied:
`Deadline.toString()` for sure is useful. But what should it look like? The problem I see with this implementation is, if someone logs a `Deadline` object, only deadlines from the same JVM process can be compared meaningfully, as the `System.nanoTime()` part of `deadlineNanos` is kinda random. 

Should I implement this to return a proper date/time instead?

carl-mastrangelo replied:
It would be nice if the units were printed out too.  We have a class to do this for timeouts in GrpcUtil that is almost what you want.

ejona86 replied:
I wouldn't mind if it was:
```java
return "" + timeRemaining(NANOSECONDS) + "ns from now";
```

GrpcUtil is another option. I think it would probably be more useful to be relative instead of printing out the date/time, but date/time would be more helpful than just the raw deadlineNanos.

carl-mastrangelo replied:
FWIW I just implemented a timer heap in a personal project of mine, and I agree with Eric: relative times are really useful, absolute times are harder to read.  Since log statements typically contain a timestamp anyways, relative doesn't seem so bad.

buchgr replied:
@carl-mastrangelo good point!

ejona86 commented:
@buchgr LGTM

ejona86 commented:
codecov/patch is non-binding. It can be okay to submit if it fails, but you should at least consider if you should add more tests. Reaching 90-95% coverage requires constant vigilance!

buchgr commented:
@carl-mastrangelo @ejona86 PTAL

carl-mastrangelo commented on core/src/test/java/io/grpc/CallOptionsTest.java:
```diff
@@ -108,27 +114,39 @@ public void mutateExecutor() {
-    assertEquals("CallOptions{deadlineNanoTime=null, authority=authority}",
-        allSet.withDeadlineNanoTime(null).toString());
-    assertTrue(allSet.toString().contains("deadlineNanoTime=" + sampleDeadlineNanoTime + ","));
+    assertEquals("CallOptions{deadline=null, authority=authority, "
```
Just check that the string .contains("123ns form now")

buchgr replied:
I do, 3 lines down :).

carl-mastrangelo replied:
Ah.  In that case this line should be changed to check for absence rather than exact string matching.

buchgr replied:
well it does check for absence by `deadline=null`?

carl-mastrangelo replied:
The idea here is to reduce burden when updating CallOptions.  If these tests spuriously fail because the string representation intended solely for debugging was slightly off, that's bad.  Checking for just the snippet you care about in the string not only solves this, but make it more explicit what the test is actually expecting.  It's a maintenance win.

carl-mastrangelo commented:
LGTM

buchgr commented on core/src/test/java/io/grpc/CallOptionsTest.java:
```diff
@@ -108,27 +118,55 @@ public void mutateExecutor() {
+    long remainingNanos = extractRemainingTime(allSet.toString());
+    long delta = TimeUnit.MILLISECONDS.toNanos(10);
+    assertNotNull(allSet.getDeadline());
+    assertEquals(remainingNanos, allSet.getDeadline().timeRemaining(NANOSECONDS), delta);
```
@carl-mastrangelo added a regex to match for `Deadline.toString` as well as filter the nanotime. WDYT?

buchgr commented:
@ejona86 @carl-mastrangelo PTAL

Should be good to go now. Modified `Deadline.timeRemaining()` to return a negtive offset in a separate commit. Added some more tests for `toString()` in `CallOptionsTest` and `DeadlineTest` that uses a regex to better check the deadline.

ejona86 commented on core/src/main/java/io/grpc/Deadline.java:
```diff
@@ -114,18 +114,15 @@ public Deadline offset(long offset, TimeUnit units) {
    * How much time is remaining in the specified time unit. Internal units are maintained as
    * nanoseconds and conversions are subject to the constraints documented for
-   * {@link TimeUnit#convert}. If there is no time remaining {@code 0} is always returned.
+   * {@link TimeUnit#convert}. If there is no time remaining, a negative offset from the
```
Mentioning "system clock" is a bit strange, although it could be fine. You might consider something like, "If there is no time remaining, the returned duration is how long ago the deadline expired."

carl-mastrangelo commented on core/src/test/java/io/grpc/CallOptionsTest.java:
```diff
@@ -108,27 +118,55 @@ public void mutateExecutor() {
+    assertTrue(opts.getDeadline().isExpired());
+  }
+
+  private static long extractRemainingTime(String deadlineStr) {
```
This seems like overengineering, but meh.

buchgr replied:
tiny bit :smiley:

ejona86 commented on core/src/test/java/io/grpc/CallOptionsTest.java:
```diff
@@ -108,27 +118,55 @@ public void mutateExecutor() {
+  }
+
+  @Test
+  @SuppressWarnings("deprecation")
```
A better approach may be to just mark the test as `@Deprecated`.

ejona86 commented:
@buchgr LGTM

buchgr commented:
thanks guys! addressed the comments. will squash and merge once CI is done.

ejona86 commented:
@buchgr, if Visual Studio 2013 passes, that's commonly "good enough" and not worth waiting for Travis. That doesn't mean ignore Travis if it fails though :smile:

buchgr commented:
cherry-picked (a4640c88ceb4777c0ba2d83eb3e2861e1da2030c) and (40575bc9fa99a8c69ad23943f287917008cd91b9).

@ejona86 I ll make sure to blame a broken build on you :stuck_out_tongue_winking_eye:
